File: src/data/4o/Maximize_the_Number_of_Partitions_After_Operations/Maximize_the_Number_of_Partitions_After_Operations.java
import java.util.*;

class Solution {
    public int maxPartitionsAfterOperations(String s, int k) {
        int n = s.length();
        int maxPartitions = 0;

        // First, calculate the number of partitions without any change
        maxPartitions = getMaxPartitions(s, k);

        // Now, try changing each character and check if it increases the number of partitions
        for (int i = 0; i < n; i++) {
            for (char c = 'a'; c <= 'z'; c++) {
                if (s.charAt(i) != c) {
                    String modified = s.substring(0, i) + c + s.substring(i + 1);
                    maxPartitions = Math.max(maxPartitions, getMaxPartitions(modified, k));
                }
            }
        }

        return maxPartitions;
    }

    private int getMaxPartitions(String s, int k) {
        int[] charCount = new int[26];
        int distinctCount = 0;
        int partitions = 0;
        int left = 0;

        for (int right = 0; right < s.length(); right++) {
            char c = s.charAt(right);
            if (charCount[c - 'a'] == 0) {
                distinctCount++;
            }
            charCount[c - 'a']++;

            while (distinctCount > k) {
                char leftChar = s.charAt(left);
                charCount[leftChar - 'a']--;
                if (charCount[leftChar - 'a'] == 0) {
                    distinctCount--;
                }
                left++;
            }

            if (distinctCount <= k) {
                partitions++;
                left = right + 1; // Move to the next partition
            }
        }

        return partitions;
    }
}

public class Maximize_the_Number_of_Partitions_After_Operations {
public static void runTests(Solution solution) {
    // Test case 1: Single character string
    String s1 = "a";
    int k1 = 1;
    int expected_result1 = 1;
    assert solution.maxPartitionsAfterOperations(s1, k1) == expected_result1 : "Test case 1 failed. Expected " + expected_result1;

    // Test case 2: No change needed to partition optimally
    String s2 = "accca";
    int k2 = 2;
    int expected_result2 = 3;
    assert solution.maxPartitionsAfterOperations(s2, k2) == expected_result2 : "Test case 2 failed. Expected " + expected_result2;

    // Test case 3: Change one character to maximize partitions
    String s3 = "abac";
    int k3 = 2;
    int expected_result3 = 2;
    assert solution.maxPartitionsAfterOperations(s3, k3) == expected_result3 : "Test case 3 failed. Expected " + expected_result3;

    // Test case 4: Entire string has more than k distinct characters, requiring changes
    String s4 = "aabaab";
    int k4 = 3;
    int expected_result4 = 1;
    assert solution.maxPartitionsAfterOperations(s4, k4) == expected_result4 : "Test case 4 failed. Expected " + expected_result4;

    // Test case 5: Optimal choice without changing characters
    String s5 = "xxyz";
    int k5 = 1;
    int expected_result5 = 4;
    assert solution.maxPartitionsAfterOperations(s5, k5) == expected_result5 : "Test case 5 failed. Expected " + expected_result5;

    // Test case 6: Large k that allows full string as one partition
    String s6 = "abcdefgh";
    int k6 = 8;
    int expected_result6 = 1;
    assert solution.maxPartitionsAfterOperations(s6, k6) == expected_result6 : "Test case 6 failed. Expected " + expected_result6;

    // Test case 7: k smaller than distinct characters in s, requiring multiple partitions
    String s7 = "abcabcabc";
    int k7 = 1;
    int expected_result7 = 9;
    assert solution.maxPartitionsAfterOperations(s7, k7) == expected_result7 : "Test case 7 failed. Expected " + expected_result7;
}


    public static void main(String[] args) {
        Solution solution = new Solution();
        for (int i = 0; i < 100000; i++) {
            runTests(solution); // Ensure your test cases are implemented inside runTests
        }
    }
}

    

File: src/data/4o/Maximize_the_Number_of_Partitions_After_Operations/Maximize_the_Number_of_Partitions_After_Operations.cpp
#include <string>
#include <unordered_map>
#include <algorithm>
using namespace std;

class Solution {
public:
    int maxPartitionsAfterOperations(string s, int k) {
        int n = s.length();
        int maxPartitions = countPartitions(s, k); // Count partitions without any change

        // Try changing each character and count the partitions
        for (int i = 0; i < n; i++) {
            char original = s[i];
            for (char c = 'a'; c <= 'z'; c++) {
                if (c != original) {
                    s[i] = c;
                    maxPartitions = max(maxPartitions, countPartitions(s, k));
                }
            }
            s[i] = original; // Restore the original character
        }

        return maxPartitions;
    }

private:
    // Helper function to count partitions with at most k distinct characters
    int countPartitions(const string& s, int k) {
        unordered_map<char, int> freq;
        int distinctCount = 0;
        int partitions = 0;

        for (char c : s) {
            freq[c]++;
            if (freq[c] == 1) distinctCount++; // New distinct character

            if (distinctCount > k) {
                partitions++;
                freq.clear(); // Reset frequency map for the new partition
                distinctCount = 1; // Current character starts the new partition
                freq[c]++;
            }
        }

        return partitions + 1; // Add the last partition
    }
};

//  Test cases for the C++ solution
#include <cassert>

// Test cases for the C++ solution
void testSolution(Solution& solution) {
    // Test case 1: Single character string
    std::string s = "a";
    int k = 1;
    int expected_result = 1;
    assert(solution.maxPartitionsAfterOperations(s, k) == expected_result && "Test case 1 failed.");

    // Test case 2: No change needed to partition optimally
    s = "accca";
    k = 2;
    expected_result = 3;
    assert(solution.maxPartitionsAfterOperations(s, k) == expected_result && "Test case 2 failed.");

    // Test case 3: Change one character to maximize partitions
    s = "abac";
    k = 2;
    expected_result = 2;
    assert(solution.maxPartitionsAfterOperations(s, k) == expected_result && "Test case 3 failed.");

    // Test case 4: Entire string has more than k distinct characters, requiring changes
    s = "aabaab";
    k = 3;
    expected_result = 1;
    assert(solution.maxPartitionsAfterOperations(s, k) == expected_result && "Test case 4 failed.");

    // Test case 5: Optimal choice without changing characters
    s = "xxyz";
    k = 1;
    expected_result = 4;
    assert(solution.maxPartitionsAfterOperations(s, k) == expected_result && "Test case 5 failed.");

    // Test case 6: Large k that allows full string as one partition
    s = "abcdefgh";
    k = 8;
    expected_result = 1;
    assert(solution.maxPartitionsAfterOperations(s, k) == expected_result && "Test case 6 failed.");

    // Test case 7: k smaller than distinct characters in s, requiring multiple partitions
    s = "abcabcabc";
    k = 1;
    expected_result = 9;
    assert(solution.maxPartitionsAfterOperations(s, k) == expected_result && "Test case 7 failed.");
}

int main() {
    Solution solution;
    for (int i = 0; i < 2000; ++i) {
        testSolution(solution);  
    }
    return 0;
}
